subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java

```diff
@@ -97,32 +101,66 @@ public FileCacheBackedScriptClassCompiler(CacheRepository cacheRepository, Scrip
         // Both caches can be closed directly after use because:
         // For 1, if the script changes or its compile classpath changes, a different directory will be used
         // For 2, if the script changes, a different cache is used. If the classpath changes, the cache is invalidated, but classes are remapped to 1. anyway so never directly used
-        PersistentCache remappedClassesCache = cacheRepository.cache("scripts-remapped/" + source.getClassName() + "/" + sourceHash + "/" + classpathHash)
-            .withDisplayName(dslId + " remapped class cache for " + sourceHash)
-            .withInitializer(new ProgressReportingInitializer(progressLoggerFactory, new RemapBuildScriptsAction<>(remapped, classpathHash, sourceHash, dslId, classLoader, operation, verifier, scriptBaseClass),
+        final PersistentCache cache = cacheRepository.cache("scripts/" + sourceHash + "/" + classpathHash)
+            .withDisplayName(dslId + " generic class cache for " + source.getDisplayName())
+            .withInitializer(new ProgressReportingInitializer(
+                progressLoggerFactory,
+                new CompileToCrossBuildCacheAction(remapped, classLoader, operation, verifier, scriptBaseClass),
                 "Compiling script into cache",
-                "Compiling " + source.getFileName() + " into local compilation cache"))
+                "Compiling " + source.getDisplayName()))
             .open();
         try {
-            File remappedClassesDir = classesDir(remappedClassesCache);
-            File remappedMetadataDir = metadataDir(remappedClassesCache);
-
-            return scriptCompilationHandler.loadFromDir(source, sourceHashCode, targetScope, remappedClassesDir, remappedMetadataDir, operation, scriptBaseClass);
+            File genericClassesDir = classesDir(cache, operation);
+            File metadataDir = metadataDir(cache);
+            File remappedClassesDir = remapClasses(genericClassesDir, remapped);
```

Comment:
@adammurdoch this line takes ~400ms to execute when 2 independent Gradle projects are being configured in parallel but normally it takes ~1ms.